src/SetupApi/SetupApi.cs

```diff
@@ -219,6 +219,272 @@ public static partial class SetupApi
+            Guid* classGuid,
+            IntPtr hwndParent);
+
+        /// <summary>
+        /// The <see cref="SetupDiOpenDeviceInfo"/> function adds a device information element for a device instance to a device information set,
+        /// if one does not already exist in the device information set, and retrieves information that identifies the device information element
+        /// for the device instance in the device information set.
+        /// </summary>
+        /// <param name="deviceInfoSet">
+        /// A handle to the device information set to which <see cref="SetupDiOpenDeviceInfo"/> adds a device information element, if one does
+        /// not already exist, for the device instance that is specified by <paramref name="deviceInstanceId"/>.
+        /// </param>
+        /// <param name="deviceInstanceId">
+        /// A <see cref="string"/> that supplies the device instance identifier of a device (for example, <c>Root\*PNP0500\0000</c>). If
+        /// <paramref name="deviceInstanceId"/> is <see langword="null"/> or <see cref="string.Empty"/>, <see cref="SetupDiOpenDeviceInfo"/> adds a
+        /// device information element to the supplied device information set, if one does not already exist, for the root device in the device tree.
+        /// </param>
+        /// <param name="parent">
+        /// The handle to the top-level window to use for any user interface related to installing the device.
+        /// A <see cref="OpenFlags"/> that controls how the device information element is opened.
+        /// </param>
+        /// <param name="openFlags">
+        /// A variable of <see cref="OpenFlags"/> that controls how the device information element is opened.
+        /// </param>
+        /// <param name="deviceInfoData">
+        /// A pointer to a caller-supplied <see cref="SP_DEVINFO_DATA"/> structure that receives information about the device information
+        /// element for the device instance that is specified by <paramref name="deviceInstanceId"/>. The caller must set <see cref="SP_DEVINFO_DATA.Size"/>.
+        /// This parameter is optional and can be <see langword="null"/>.
+        /// </param>
+        /// <returns>
+        /// The function returns <see langword="true"/> if it is successful.
+        /// Otherwise, it returns <see langword="false"/> and the logged error can be retrieved with a call to <see cref="Marshal.GetLastWin32Error"/>.
+        /// </returns>
+        /// <see href="https://msdn.microsoft.com/en-us/library/windows/hardware/ff552071(v=vs.85).aspx"/>
+        [DllImport(nameof(SetupApi), SetLastError = true)]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        public static extern bool SetupDiOpenDeviceInfo(
+            SafeDeviceInfoSetHandle deviceInfoSet,
+            string deviceInstanceId,
+            IntPtr parent,
+            OpenFlags openFlags,
+            ref SP_DEVINFO_DATA deviceInfoData);
```

Comment:
The doc indicates this last parameter is optional and can be null. Can you please change this parameter to:
```cs
[Friendly(FriendlyFlags.Bidirectional | FriendlyFlags.Optional)] SP_DEVINFO_DATA*
```
I believe we'll still generate a `ref Struct` overload, but folks can then also pass `null` as an argument instead.
BTW, this won't work if `SP_DEVINFO_DATA` has `MarshalAs` attributed fields on it, which is why we also discourage those attributes in structs, also mentioned in our contributing doc. :)

Reply:
The parameter is not optional, so I've changed the argument to `[Friendly(FriendlyFlags.Bidirectional)] SP_DEVINFO_DATA*`